shell/platform/darwin/macos/framework/Source/FLEExternalTextureGL.mm

```diff
@@ -0,0 +1,89 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "flutter/shell/platform/darwin/macos/framework/Source/FLEExternalTextureGL.h"
+
+#import <AppKit/AppKit.h>
+#import <CoreVideo/CoreVideo.h>
+#import <OpenGL/gl.h>
+
+static void OnGLTextureRelease(CVPixelBufferRef pixelBuffer) {
+  CVPixelBufferRelease(pixelBuffer);
+}
+
+@implementation FLEExternalTextureGL {
+  /**
+   * OpenGL texture cache.
+   */
+  CVOpenGLTextureCacheRef _openGLTextureCache;
+  /**
+   * The pixel buffer copied from the user side will be released
+   * when the flutter engine renders it.
+   */
+  CVPixelBufferRef _pixelBuffer;
+  /**
+   * User side texture object, used to copy pixel buffer.
+   */
+  id<FLETexture> _texture;
+}
+
+- (instancetype)initWithFLETexture:(id<FLETexture>)texture {
+  self = [super init];
+  if (self) {
+    _texture = texture;
+  }
+  return self;
+}
+
+- (int64_t)textureID {
+  return reinterpret_cast<int64_t>(self);
+}
+
+- (BOOL)populateTextureWithWidth:(size_t)width
+                          height:(size_t)height
+                         openGLTexture:(FlutterOpenGLTexture*)openGLTexture {
+  // Copy the pixel buffer from the FLETexture instance implemented on the user side.
+  _pixelBuffer = [_texture copyPixelBuffer:width height:height];
+
+  if (!_pixelBuffer) {
+    return NO;
+  }
+
+  // Create the opengl texture cache if necessary.
+  if (!_openGLTextureCache) {
+    CGLContextObj context = [NSOpenGLContext currentContext].CGLContextObj;
+    CGLPixelFormatObj format = CGLGetPixelFormat(context);
+    if (CVOpenGLTextureCacheCreate(kCFAllocatorDefault, NULL, context, format, NULL,
+                                   &_openGLTextureCache) != kCVReturnSuccess) {
+      NSLog(@"Could not create texture cache.");
+      CVPixelBufferRelease(_pixelBuffer);
+      return NO;
+    }
+  }
+
+  // Try to clear the cache of OpenGL textures to save memory.
+  CVOpenGLTextureCacheFlush(_openGLTextureCache, 0);
+
+  CVOpenGLTextureRef cvOpenGLTexture = NULL;
+  if (CVOpenGLTextureCacheCreateTextureFromImage(kCFAllocatorDefault, _openGLTextureCache,
+                                                 _pixelBuffer, NULL,
+                                                 &cvOpenGLTexture) != kCVReturnSuccess) {
+    CVPixelBufferRelease(_pixelBuffer);
+    return NO;
+  }
+  CVPixelBufferRelease(_pixelBuffer);
+
+  openGLTexture->target = static_cast<uint32_t>(CVOpenGLTextureGetTarget(cvOpenGLTexture));
```

Comment:
@cloudwebrtc Are you saying that  CVOpenGLTextureGetTarget is not returning GL_TEXTURE_RECTANGLE? I am a little bit confused. There should be no issues with having to convert to GL_TEXTURE_2D here.

Reply:
@chinmaygarde 
The whole picture of the problem is:
- 1, when I patched Texture for flutter/engine to macOS 10.14.4 (xcode 10.2)
The following error will appear in the console at runtime, without displaying any texture frames.
![image](https://user-images.githubusercontent.com/17733746/56361486-0ab7cc00-621a-11e9-8146-c47d8b600fd7.png)
- 2, so I tried to find the reason
I solved it by patching skia
``` patch
diff --git a/src/sksl/SkSLGLSLCodeGenerator.cpp b/src/sksl/SkSLGLSLCodeGenerator.cpp
index d003dc9397..04357e2a4c 100644
--- a/src/sksl/SkSLGLSLCodeGenerator.cpp
+++ b/src/sksl/SkSLGLSLCodeGenerator.cpp
@@ -656,7 +656,7 @@ void GLSLCodeGenerator::writeFunctionCall(const FunctionCall& c) {
                         proj = false;
                         break;
                     case SpvDimRect:
-                        dim = "Rect";
+                        dim = "2DRect";
                         proj = false;
                         break;
                     case SpvDimBuffer:
```
Then I can see the texture display, but the picture is not correct.
![image](https://user-images.githubusercontent.com/17733746/56362684-0e008700-621d-11e9-93fb-688c88aaca2a.png)
- 3, then I use the width/height in TextureFrameCallback to resized the video frame to this size on the WebRTC side, then get the texture display of the correct zoom size, and feel free to stretch ExampleEmbedder.app to display correctly.
[shell/platform/embedder/embedder.h#L214](https://github.com/flutter/engine/blob/master/shell/platform/embedder/embedder.h#L214)
``` cpp
Typedef bool (*TextureFrameCallback)(void* /* user data */,
                                      Int64_t /* texture identifier */,
                                      Size_t /* width */,
                                      Size_t /* height */,
                                      FlutterOpenGLTexture* /* texture out */);
```
Looks like this:
[plugins/flutter_webrtc/macos/FlutterRTCVideoRenderer.m#L52](https://github.com/cloudwebrtc/flutter-desktop-embedding/blob/test_for_new_macos_texture/plugins/flutter_webrtc/macos/FlutterRTCVideoRenderer.m#L52)

I am not sure that the type of dim in SpvDimRect is a bug in skia on macos.